Incorrect Currency Displayed When Order Currency Differs from Store Default

AnirudhBhat opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #12865 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR disables `Edit` option for orders whose currency differs from store currency.

**Reason**
We consistently use the store's default currency when editing an order, even if the order was created in a different currency. This results in inconsistencies across various sections of the editable order screen, including Items, Custom Amounts, Discounts, Coupons, and Order Totals. Additionally, when adding new products to the order, prices are displayed in the store's default currency instead of the order's currency.

This inconsistency is confusing for users.

**Suggested Workaround to consider:**

> To completely fix this issue, we would need to receive product prices in the order’s currency so we can display them correctly during product selection. Since the REST API does not currently provide this, I suggest disabling the Edit mode for orders in a currency different from the store’s default.
> 
> Users would still be able to change the order status, create shipping labels, issue refunds, and add tracking, notes, and shipping details from the Order Details. They wouldn’t be able to add Custom Amounts, Discounts, Coupons, or new Items to the order. These actions aren’t very common for orders created on the storefront anyways.

**More context**
peaMlT-XF-p2


### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
1. In your store viewfront, create an order with some items with a currency other than the default one. You can do that by accessing the store from a country with a different currency than the default one (better in Incognito mode in the browser). For that to work, you should have enabled the Automatically switch customers to their local currency if it has been enabled in the Multicurrency setting of WooCommerce wp-admin.
2. Open the app and go to orders.
3. Click on Edit orders
4. See that the currency displayed is not proper.

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
Tested that edit option is disabled for orders which have different currency than store currency and enabled for those whose currency matches with that of stores.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/30ebd308-4123-4798-a49d-60ebd520028a



- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

AnirudhBhat committed:
Add code to check if order currency is same as store.

AnirudhBhat committed:
Enable or disable edit menu option based on order and store currencies equality.

AnirudhBhat committed:
Add test to verify edit menu is disabled when order currency is different than store

AnirudhBhat committed:
Add test to verify edit menu is enabled when order currency is same as store

AnirudhBhat committed:
Move string parameters_key into private constant.

AnirudhBhat committed:
Fix detekt errors

AnirudhBhat committed:
Add release notes

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  1 Message: <p>This PR contains changes to...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/243">21.1</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false"><p>This PR contains changes to Tracks-related logic. Please ensure (<strong>author and reviewer</strong>) the following are completed:</p>

<ul>
  <li>The tracks events must be validated in the Tracks system.</li>
  <li>Verify the internal Tracks spreadsheet has also been updated.</li>
  <li>Please consider registering any new events.</li>
  <li>The PR must be assigned the <strong>category: tracks</strong> label.</li>
</ul></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt:
```diff
@@ -261,6 +263,13 @@ class OrderDetailViewModel @Inject constructor(
     fun hasOrder() = viewState.orderInfo?.order != null
 
+    private fun getStoreCurrency(): String {
+        return parameterRepository.getParameters(
```
Aside from the questionability of the API of this repository, I'm not sure I understand why you passed there `savedState`? How can it contain a value with key `parameters_key` if nobody puts it there? Unless I am missing something, it just always calls `loadParameters` inside of it.

Having said that, I'd recommend not using this repo at all:
* It's located in the product package
* This API is questionable, so I'd say refactoring is needed. Less usage -> easier to refactor eventually
* It hits the database from the main thread now

You can simply call

```
wooCommerceStore.getSiteSettings(selectedSite.get())?.currencyCode
```

Just make sure to not call it from the main thread

AnirudhBhat replied:
I agree and I have refactored it by moving the logic into a separate class.

kidinov commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/OrderDetailViewModelTest.kt:
```diff
@@ -2427,4 +2433,64 @@ class OrderDetailViewModelTest : BaseUnitTest() {
+    fun `when order currency is different than store, then disable edit menu in order creation`() = testBlocking {
+        val nonRefundedOrder = order.copy(refundTotal = BigDecimal.ZERO)
+
+        doReturn(false).whenever(paymentCollectibilityChecker).isCollectable(any())
```
I believe that in most cases, we should continue to use the "whenever" syntax for mocking. Although this test was written differently (even though the initial tests were implemented that way), I suggest that we adhere to the project's coding style. Ideally, we should consider rewriting the other tests to align with the syntax predominantly used in the project, possibly with the help of AI.

AnirudhBhat replied:
I've updated the new tests with the "whenever" syntax and I'll update the remaining tests with the following PR.

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt:
```diff
@@ -317,6 +326,14 @@ class OrderDetailViewModel @Inject constructor(
         )
     }
 
+    fun isOrderCurrencySameAsStore(): Boolean {
```
I'd actually go a step further here, in order to avoid enormous-sized VMs, e.g.:

```
class IsStoreCurrencyMatch(private val wooCommerceStore: WooCommerceStore) {
    
    suspend operator fun invoke(): Boolean {
        val storeCurrency = withContext(Dispatchers.IO) {
            wooCommerceStore.getSiteSettings()?.currencyCode
        }
        return storeCurrency.equals(wooCommerceStore.getOrderCurrency(), ignoreCase = true)
    }
}
```
```
if (isStoreCurrencyMatch(order.currency)) {...
```

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt:
```diff
@@ -317,6 +326,14 @@ class OrderDetailViewModel @Inject constructor(
+        return if (hasOrder()) {
+            getStoreCurrency().equals(order.currency, ignoreCase = true)
+        } else {
+            true
```
Why is this true if it doesn't have an order?

kidinov reviewed (commented):
As discussed on slack a small change regarding the way how we handling it UX wise would be nice to add.

I added a few comments regarding my vision how we can improve the code, please let me know wdyt?

AnirudhBhat committed:
Add a new class that detects whether store currency is different than order currency

AnirudhBhat committed:
Add string resource to display when store and order currency are different

AnirudhBhat committed:
Make use of IsStoreCurrencyMisMatch class in view model

AnirudhBhat committed:
Remove unnecessary code

AnirudhBhat committed:
Remove unnecessary tests

AnirudhBhat committed:
Add test to verify correct data is returned when there is a mismatch

AnirudhBhat committed:
Add test to verify correct data is returned when there is a match

AnirudhBhat committed:
Remove unused imports

AnirudhBhat committed:
Add test to verify show snackbar event is triggered when there is order and store currency mismatch

AnirudhBhat committed:
Add test to verify show EditOrder event is triggered when there is order and store currency are same

AnirudhBhat committed:
Fix detekt errors

AnirudhBhat committed:
Add analytics event to track currency conflict

AnirudhBhat committed:
Add a method to track currency conflict

AnirudhBhat committed:
Track currency conflict in view model when edit button is clicked

AnirudhBhat committed:
Add test to verify currency conflict event is tracked

AnirudhBhat committed:
Merge branch 'refs/heads/trunk' into issue/12865-incorrect-currency-displayed

AnirudhBhat commented:
@kidinov Thanks for all the suggestions 🙏 I've refactored it and have implemented the snack bar display as we discussed on Slack. Could you please take another look?

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/IsStoreCurrencyMissMatch.kt:
```diff
@@ -0,0 +1,28 @@
+import org.wordpress.android.fluxc.store.WooCommerceStore
+import javax.inject.Inject
+
+class IsStoreCurrencyMatch @Inject constructor(
```
The file name doesn't match the class name. As you return `isMatch` then it sounds like the file should be `IsStoreCurrencyMatch` or `IsStoreCurrencyMatches`

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12880/commits/c275d245bfdb473ba4a0ff8b08ed191839172a0c

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailFragment.kt:
```diff
@@ -164,6 +164,7 @@ class OrderDetailFragment :
 
     override fun onResume() {
         super.onResume()
+        onPrepareMenu(binding.toolbar.menu)
```
Is this needed?

AnirudhBhat replied:
Good catch! Not needed. Removed here: https://github.com/woocommerce/woocommerce-android/pull/12880/commits/7241eb27d40c76b7d47df5c3ac284609dd7e2dc9

AnirudhBhat committed:
Remove unnecessary code

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt:
```diff
@@ -371,15 +373,28 @@ class OrderDetailViewModel @Inject constructor(
+        launch {
+            val isCurrencyMatch = isStoreCurrencyMatch(order.currency)
+            if (!isCurrencyMatch.isMatch) {
+                tracker.trackOrderAndStoreCurrencyMismatchWhenEditButtonTapped()
```
np: I'd suggest extracting both branches, for better readability, e.g.:
* `handleEditClickedWithCurrencyMissmatch`
* `handleEditClicked`
or something like that

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12880/commits/2fc8aa3020356a775a33ccc9d40b9b6bea7dc54a

AnirudhBhat committed:
Rename file so that it would match class name

kidinov commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/IsStoreCurrencyMisMatchTest.kt:
```diff
@@ -0,0 +1,40 @@
+    private val wooCommerceStore: WooCommerceStore = mock()
+    private val selectedSite: SelectedSite = mock()
+    val wcSettingsModel = mock<WCSettingsModel> {
+        on { currencyCode } doReturn "USD"
```
np: Not sure about that but I think usually we use `on { currencyCode }.thenReturn("USD")` syntax?

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12880/commits/7ddc07a0cd60d4d8270db0d43e682305a78a291b

AnirudhBhat committed:
Extract method for readability

AnirudhBhat committed:
Use thenReturn instead of doReturn

kidinov commented:
@AnirudhBhat 

it crashes for me with on a 2 pane layout 🤔 . ~Is this because of `onPrepareMenu`?~ - nope

```
Process: com.woocommerce.android.dev, PID: 5864
java.lang.IllegalStateException: Multiple observers registered but only one is supported.
                                                                                                    	at com.woocommerce.android.viewmodel.LiveDataDelegate.observe(LiveDataDelegate.kt:53)
                                                                                                    	at com.woocommerce.android.ui.orders.creation.OrderCreateEditFormFragment.observeViewStateChanges(OrderCreateEditFormFragment.kt:486)
                                                                                                    	at com.woocommerce.android.ui.orders.creation.OrderCreateEditFormFragment.setupObserversWith(OrderCreateEditFormFragment.kt:417)
                                                                                                    	at com.woocommerce.android.ui.orders.creation.OrderCreateEditFormFragment.onViewCreated(OrderCreateEditFormFragment.kt:164)
                                                                                                    	at androidx.fragment.app.Fragment.performViewCreated(Fragment.java:3152)
                                                                                                    	at androidx.fragment.app.FragmentStateManager.createView(FragmentStateManager.java:608)
                                                                                                    	at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:286)
                                                                                                    	at androidx.fragment.app.FragmentManager.executeOpsTogether(FragmentManager.java:2164)
                                                                                                    	at androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute(FragmentManager.java:2065)
                                                                                                    	at androidx.fragment.app.FragmentManager.execPendingActions(FragmentManager.java:2002)
                                                                                                    	at androidx.fragment.app.FragmentManager$5.run(FragmentManager.java:702)
                                                                                                    	at android.os.Handler.handleCallback(Handler.java:942)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:99)
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:201)
                                                                                                    	at android.os.Looper.loop(Looper.java:288)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:7924)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
```

AnirudhBhat committed:
Fix failing test

kidinov reviewed (commented):
@AnirudhBhat  👋

The case with currency miss match works fine, but it crashes when a currency matches and i could not tell from the first sight why

I left a couple of minor comments too, please take a look

AnirudhBhat committed:
Remove unused import

AnirudhBhat commented:
>it crashes for me with on a 2 pane layout 

It seems like the crash is unrelated to this change. The crash happens even on latest `trunk`. Raised this on our internal Slack channel.

I've addressed your feedback. Thanks!

kidinov reviewed (approved):
Thanks @AnirudhBhat ! LGTM

The only part which is not ideal in my eyes is that there are cases when we show empty string instead of currency to users, but we can not do anything here
